crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -505,6 +473,83 @@ fn handle_own_line_comment_after_branch<'a>(
     }
 }
 
+/// Handle parenthesized comments. A parenthesized comment is a comment that appears within a
+/// parenthesis, but not within the range of the expression enclosed by the parenthesis.
+/// For example, the comment here is a parenthesized comment:
+/// ```python
+/// if (
+///     # comment
+///     True
+/// ):
+///     ...
+/// ```
+/// The parentheses enclose `True`, but the range of `True`doesn't include the `# comment`.
+///
+/// Default handling can get parenthesized comments wrong in a number of ways. For example, the
+/// comment here is marked (by default) as a trailing comment of `x`, when it should be a leading
+/// comment of `y`:
+/// ```python
+/// assert (
+///     x
+/// ), ( # comment
+///     y
+/// )
+/// ```
+///
+/// Similarly, this is marked as a leading comment of `y`, when it should be a trailing comment of
+/// `x`:
+/// ```python
+/// if (
+///     x
+///     # comment
+/// ):
+///    y
+/// ```
+///
+/// As a generalized solution, if a comment has a preceding node and a following node, we search for
+/// opening and closing parentheses between the two nodes. If we find a closing parenthesis between
+/// the preceding node and the comment, then the comment is a trailing comment of the preceding
+/// node. If we find an opening parenthesis between the comment and the following node, then the
+/// comment is a leading comment of the following node.
+fn handle_parenthesized_comment<'a>(
+    comment: DecoratedComment<'a>,
+    locator: &Locator,
+) -> CommentPlacement<'a> {
+    let Some(preceding) = comment.preceding_node() else {
+        return CommentPlacement::Default(comment);
+    };
+
+    let Some(following) = comment.following_node() else {
+        return CommentPlacement::Default(comment);
+    };
+
+    if comment.line_position().is_end_of_line() {
+        let tokenizer = SimpleTokenizer::new(
+            locator.contents(),
+            TextRange::new(preceding.end(), comment.start()),
+        );
+        if tokenizer
+            .skip_trivia()
+            .any(|token| token.kind() == SimpleTokenKind::LParen)
```

Comment:
Is there a case where the range may contain any tokens that the `SimpleTokenizer` doesn't understand and makes it return `Bogus` tokens? If so, how do we ensure that the formatting is consistent between cases where all tokens are supported by the `SimpleTokenizer` and cases where they are not? Should we bail as soon as we see a `Bogus` token (or add a debug assertion that we never see a bogus token?).

Same applies for the trailing lexing

Reply:
I don't believe so -- this is why I needed to extend the `SimpleTokenizer` to support all trivia that can appear between AST nodes. If it does return `Bogus`, that's a bug in the tokenizer. I'll add a debug assertion to that effect.